File: src/model/instrument/cross_section_lut.zig
const std = @import("std");
const errors = @import("../../core/errors.zig");
const LutControls = @import("../../core/lut_controls.zig");
const build_helpers = @import("cross_section_lut_build.zig");
const eval_helpers = @import("cross_section_lut_eval.zig");
const constants = @import("constants.zig");

const Allocator = std.mem.Allocator;
const max_operational_refspec_temperature_coefficients = constants.max_operational_refspec_temperature_coefficients;
const max_operational_refspec_pressure_coefficients = constants.max_operational_refspec_pressure_coefficients;

pub const GenerationSource = build_helpers.GenerationSource;

pub const OperationalCrossSectionLut = struct {
    wavelengths_nm: []const f64 = &[_]f64{},
    coefficients: []const f64 = &[_]f64{},
    temperature_coefficient_count: u8 = 0,
    pressure_coefficient_count: u8 = 0,
    min_temperature_k: f64 = 0.0,
    max_temperature_k: f64 = 0.0,
    min_pressure_hpa: f64 = 0.0,
    max_pressure_hpa: f64 = 0.0,

    pub fn enabled(self: *const OperationalCrossSectionLut) bool {
        return self.wavelengths_nm.len > 0;
    }

    pub fn validate(self: *const OperationalCrossSectionLut) errors.Error!void {
        if (!self.enabled()) {
            if (self.coefficients.len != 0 or
                self.temperature_coefficient_count != 0 or
                self.pressure_coefficient_count != 0)
            {
                return errors.Error.InvalidRequest;
            }
            return;
        }

        if (self.temperature_coefficient_count == 0 or
            self.temperature_coefficient_count > max_operational_refspec_temperature_coefficients or
            self.pressure_coefficient_count == 0 or
            self.pressure_coefficient_count > max_operational_refspec_pressure_coefficients)
        {
            return errors.Error.InvalidRequest;
        }

        if (!std.math.isFinite(self.min_temperature_k) or
            !std.math.isFinite(self.max_temperature_k) or
            !std.math.isFinite(self.min_pressure_hpa) or
            !std.math.isFinite(self.max_pressure_hpa) or
            self.min_temperature_k <= 0.0 or
            self.max_temperature_k <= self.min_temperature_k or
            self.min_pressure_hpa <= 0.0 or
            self.max_pressure_hpa <= self.min_pressure_hpa)
        {
            return errors.Error.InvalidRequest;
        }

        var previous_wavelength: ?f64 = null;
        for (self.wavelengths_nm) |wavelength_nm| {
            if (!std.math.isFinite(wavelength_nm)) return errors.Error.InvalidRequest;
            if (previous_wavelength) |previous| {
                if (wavelength_nm <= previous) return errors.Error.InvalidRequest;
            }
            previous_wavelength = wavelength_nm;
        }

        const expected_coefficient_count =
            self.wavelengths_nm.len *
            @as(usize, self.temperature_coefficient_count) *
            @as(usize, self.pressure_coefficient_count);
        if (self.coefficients.len != expected_coefficient_count) {
            return errors.Error.InvalidRequest;
        }

        for (self.coefficients) |coefficient| {
            if (!std.math.isFinite(coefficient)) return errors.Error.InvalidRequest;
        }
    }

    pub fn clone(self: OperationalCrossSectionLut, allocator: Allocator) !OperationalCrossSectionLut {
        var cloned: OperationalCrossSectionLut = .{
            .temperature_coefficient_count = self.temperature_coefficient_count,
            .pressure_coefficient_count = self.pressure_coefficient_count,
            .min_temperature_k = self.min_temperature_k,
            .max_temperature_k = self.max_temperature_k,
            .min_pressure_hpa = self.min_pressure_hpa,
            .max_pressure_hpa = self.max_pressure_hpa,
        };
        cloned.wavelengths_nm = try allocator.dupe(f64, self.wavelengths_nm);
        errdefer allocator.free(cloned.wavelengths_nm);
        cloned.coefficients = try allocator.dupe(f64, self.coefficients);
        return cloned;
    }

    pub fn buildFromSource(
        allocator: Allocator,
        wavelengths_nm: []const f64,
        source: GenerationSource,
        controls: LutControls.XsecControls,
    ) !OperationalCrossSectionLut {
        return build_helpers.buildFromSource(
            OperationalCrossSectionLut,
            allocator,
            wavelengths_nm,
            source,
            controls,
        );
    }

    pub fn deinitOwned(self: *OperationalCrossSectionLut, allocator: Allocator) void {
        allocator.free(self.wavelengths_nm);
        allocator.free(self.coefficients);
        self.* = .{};
    }

    pub fn sigmaAt(
        self: *const OperationalCrossSectionLut,
        wavelength_nm: f64,
        temperature_k: f64,
        pressure_hpa: f64,
    ) f64 {
        return eval_helpers.evaluate(@This(), self, wavelength_nm, temperature_k, pressure_hpa).sigma;
    }

    pub fn dSigmaDTemperatureAt(
        self: *const OperationalCrossSectionLut,
        wavelength_nm: f64,
        temperature_k: f64,
        pressure_hpa: f64,
    ) f64 {
        return eval_helpers.evaluate(@This(), self, wavelength_nm, temperature_k, pressure_hpa).d_sigma_d_temperature;
    }
};

File: rust_src/input/instrument/cross_section_lut.rs
use crate::{
    common::errors,
    input::instrument::{
        constants::{
            MAX_OPERATIONAL_REFSPEC_PRESSURE_COEFFICIENTS,
            MAX_OPERATIONAL_REFSPEC_TEMPERATURE_COEFFICIENTS,
        },
        cross_section_lut_eval,
    },
};

#[derive(Debug, Default, Clone, PartialEq)]
pub struct OperationalCrossSectionLut {
    pub wavelengths_nm: Vec<f64>,
    pub coefficients: Vec<f64>,
    pub temperature_coefficient_count: u8,
    pub pressure_coefficient_count: u8,
    pub min_temperature_k: f64,
    pub max_temperature_k: f64,
    pub min_pressure_hpa: f64,
    pub max_pressure_hpa: f64,
}

impl OperationalCrossSectionLut {
    pub fn enabled(&self) -> bool {
        !self.wavelengths_nm.is_empty()
    }

    pub fn validate(&self) -> Result<(), errors::Error> {
        if !self.enabled() {
            if self.coefficients.is_empty()
                && self.temperature_coefficient_count == 0
                && self.pressure_coefficient_count == 0
            {
                return Ok(());
            }
            return Err(errors::Error::InvalidRequest);
        }

        let temperature_count = usize::from(self.temperature_coefficient_count);
        let pressure_count = usize::from(self.pressure_coefficient_count);
        if temperature_count == 0
            || temperature_count > MAX_OPERATIONAL_REFSPEC_TEMPERATURE_COEFFICIENTS
            || pressure_count == 0
            || pressure_count > MAX_OPERATIONAL_REFSPEC_PRESSURE_COEFFICIENTS
        {
            return Err(errors::Error::InvalidRequest);
        }

        if !self.min_temperature_k.is_finite()
            || !self.max_temperature_k.is_finite()
            || !self.min_pressure_hpa.is_finite()
            || !self.max_pressure_hpa.is_finite()
            || self.min_temperature_k <= 0.0
            || self.max_temperature_k <= self.min_temperature_k
            || self.min_pressure_hpa <= 0.0
            || self.max_pressure_hpa <= self.min_pressure_hpa
        {
            return Err(errors::Error::InvalidRequest);
        }

        let mut previous_wavelength = None;
        for &wavelength_nm in &self.wavelengths_nm {
            if !wavelength_nm.is_finite() {
                return Err(errors::Error::InvalidRequest);
            }
            if previous_wavelength.is_some_and(|previous| wavelength_nm <= previous) {
                return Err(errors::Error::InvalidRequest);
            }
            previous_wavelength = Some(wavelength_nm);
        }

        let expected_count = self.wavelengths_nm.len() * temperature_count * pressure_count;
        if self.coefficients.len() != expected_count {
            return Err(errors::Error::InvalidRequest);
        }
        if self.coefficients.iter().any(|value| !value.is_finite()) {
            return Err(errors::Error::InvalidRequest);
        }
        Ok(())
    }

    pub fn sigma_at(&self, wavelength_nm: f64, temperature_k: f64, pressure_hpa: f64) -> f64 {
        cross_section_lut_eval::evaluate(self, wavelength_nm, temperature_k, pressure_hpa).sigma
    }

    pub fn d_sigma_d_temperature_at(
        &self,
        wavelength_nm: f64,
        temperature_k: f64,
        pressure_hpa: f64,
    ) -> f64 {
        cross_section_lut_eval::evaluate(self, wavelength_nm, temperature_k, pressure_hpa)
            .d_sigma_d_temperature
    }
}
